libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/gql.kt

```diff
@@ -1493,14 +1494,94 @@ private fun List<GQLArgument>.writeArguments(writer: SDLWriter) {
   join(writer, prefix = "(", separator = ", ", postfix = ")")
 }
 
-class GQLField(
+
+@ApolloExperimental
+sealed interface GQLNullability : GQLNode
+
+@ApolloExperimental
+class GQLRequired(override val sourceLocation: SourceLocation? = null) : GQLNullability {
```

Comment:
This is very much nitpicking but would `GQLNonNullable` / `GQLNullable` be more consistent with `GQLNullability`, and the terms in the spec (well the spec as it is currently!)

Reply:
The language is a bit mixed in the RFC currently. There is also required/optional [here](https://github.com/graphql/graphql-wg/blob/main/rfcs/ClientControlledNullability.md#definitions)

I like required/optional. An `optional` field has a nullable designator. 
I like that I can use 2 different words for why a field type is nullable.

Reply:
I've been staring at this for the last 15min and I'm not sold on anything yet. We currently have:

* designator
* nullability
* requiredStatus

I'm currently thinking of something like `modifers`:

```
sealed interface GQLTypeModifier : GQLNode
class GQLListTypeModifier : GQLTypeModifier
class GQLRequiredTypeModifier : GQLTypeModifier
class GQLOptionalTypeModifier : GQLTypeModifier
```

I like that this maps with types because at this end of the day, the `GQLType` is the source of truth for most (all?) matters and what we're effectively doing is `val finalType = field.type.applyModifiers(field.typeModifier)`

I updated this PR to track the current state of the grammar. Let's bike shed this in an upcoming meeting?